Make several tweaks to get the project to compile and run on windows (Visual Studio 15)

Thundzz opened the pull request:
Hello Benny,

I have tried to compile and run this new project on my windows system and have encountered a few problems, so I figured I might as well volunteer to solve the problems, so I went ahead and did that, and here is the detailed explanation of every change in this contribution:

- I had a problem related to the default directory structure of SDL2: When I installed SDL2 on my systemn, the default structure  was something like `%SDL2_ROOT_DIR%/include/*.h`. I need to move everything under `%SDL2_ROOT_DIR%/include/SDL2/*.h` for everything to work smoothly. I thought it may help someone so I went ahead and added a warning about that in the README.

- I moved the initialisation of static const floats in `genericMath.cpp` file rather than having them in the hpp file, which would not let me compile. (https://stackoverflow.com/a/2454082/3909725)

- Under `indexedModel.cpp` in the `createVertexArray` method, some element of the elements Array was empty, which made the program crash at runtime. I added an if statement to eliminate those empty elements. I don't know if that was the right solution to the problem, but hey, at least it made it work.

- `__attribute__ ((always_inline))` does not work and results in horrible unreadable compilation errors under **MSVC**, I added a small test to only set FORCEINLINE to this macro on GNU based systems, and just default to simple inline on all other cases.

- I could not get the program to compile with the "normal" `main` prototype on my windows machine; seemed like the only way to get it to work was by replacing `main` with `wmain`. I added a snippet of preprocessor code to choose the proper declaration at compile time. I know it's kind of a hack, but I couldn't find a better way to get it to work, and at least it's generic.

- In genericMemory.hpp, the memswap function used  to declare a  char `temp_data[size]` which would not compile on my system because the `size` argument is dynamic and only known at runtime. I replaced the `temp_data` array with a dynamic array which is destroyed at the end of the function call.

- The swizzle method in `sseVecmath.hpp` would not compile on my system, with the compiler raising a "C20557: expected constant expression" error. Since the method is not used for the moment and since I could not figure out the reason for this error, **I just commented it for the moment, hoping that you would know how to solve the issue :)**

With these few tweaks, I was able to run the project using MSVC and Visual Studio Community 15 2017.
I hope I have not broken anything for the other environments/platforms, and I hope you'll be able to integrate my fixes to the project.

Please feel free to tell me if anything looks off or needs to be changed around a bit before integrating to the project.

Sincerely,

Thundzz committed:
Make several tweaks to get the project to compile and run on windows

Thundzz committed:
Removed CONSTEXPR declaration on the static math constants in the genericMath module.

Thundzz commented:
After checking myself on a linux machine, I realized my modifications on the `genericMath` module had broken the compilation when the `CONSTEXPR` macro is actually defined and set to `constexpr`. 

In order to make the project compilable with msvc, I have deleted the `CONSTEXPR` in front of the math constant declarations in that module. I think that shouldn't be a big deal, but I'm looking forward to knowing your opinion.

BennyQBD commented:
Hey Thundzz, thanks for the pull request! It's much appreciated. You've discovered a lot of good things here.

That said, I've elected to fix these issues (except one!) slightly differently than how you've proposed, so the changes are in a new commit rather than a pull request merge.

Most importantly, note that I did NOT implement the wmain change because while that may work and it's good to know that works, it's almost certainly the wrong solution to whatever issue you had. I'd appreciate it if you could try running this again and tell me more about the issue with the regular main method, the more details, the better.

As for my alternate solutions, here's specifically what I've done:

- SDL2 Includes: Although I think it's rather strange that SDL2 didn't put it's includes in the SDL2 directory, I added the note to the readme anyways. Good catch!

- Generic Math Constants: This is the second time someone's had an issue with the way I did the GenericMath constants, so I've simply taken them out and replaced them with macros. Much simpler, much fewer places for something to go wrong.

- createVertexArray: Turns out, this was the result of a subtle bug in the way I implemented that method. It was only supposed to iterate through the loop once for each vertex element, not once for every element in the array (because that array includes both vertex elements AND elements of instanced data, hence the "0" issue because I hadn't specifed any instanced data but the empty array was still there). All fixed, shouldn't cause any more issues.

- FORCEINLINE: Good catch! This is actually something I'd meant to do, but had forgotten about. On windows, there is a keyword for this, __forceinline, so I've taken your solution and used that instead to still get the proper effect.

- GenericMemory Swap: Another good catch, but since I'd prefer not to needlessly allocate and deallocate memory in such a simple helper method, I opted to fix this a little differently and swap one element at a time for larger arrays. That way no dynamic memory allocation is needed.

- Vector Swizzle: Another good catch on the swizzle method! That method is something I'd done early on before I realized that SSE intrinsics won't let you wrap swizzling in a method. I was instead using macros (which you can see at the top of the file), but had forgotten to remove the original method (hence why it wasn't used anywhere). It's now gone, thanks for pointing it out!

Thanks again for all your help, I very much appreciate all of this! If you're willing, I'd appreciate it if you could try the new version and see if this works for you. I suspect whatever issue can be dodged by using wmain is still there, but hopefully everything else should be fixed.

Thundzz commented:
Hello Benny,

I'm happy that you found my contribution useful! The ways you decided to fix the problems I addressed in my pull request are indeed more elegant than the way I suggested!

I have checked out the new code and it seems that your changes mostly fix the problems I had. 

There is still a small issue with the platform detection and `FORCEINLINE` thing. The `OPERATING_SYSTEM_WINDOWS` macro defined (or not defined) in `platform.hpp` is used in `common/common.hpp` to decide which keyword to use for FORCEINLINE. But since `platform/platform.hpp` is not included in `common/common.hpp` the macro is never defined, and the linux version is always used.

In `common/common.hpp`, I just added the following to fix the problem:
```c++
#include "platform/platform.hpp"
```

## The "main function" saga

As for the `main` thing, as you suspected the problem is still there. The original error I get when compiling is this:

![link_error](https://user-images.githubusercontent.com/6213717/39131376-8e8a2cd4-470f-11e8-8bd8-6a3a6e6f2bb2.JPG)


| Severity |   Code  |                                                  Description                                                  | Project |                              File                             | Line | Suppression State |
|----------|---------|---------------------------------------------------------------------------------------------------------------|---------|---------------------------------------------------------------|------|-------------------|
| Error    | LNK2019 | unresolved external symbol _main referenced in function "int __cdecl invoke_main(void)" (?invoke_main@@YAHXZ) | CGFX5   | C:\Users\thundzz\Desktop\CGFX5\build\MSVCRTD.lib(exe_main.obj) |    1 |                   |


It seemed like the root cause of this problem is related to the redinition of the `main` function in the SDL library. Apparently, SDL redefines the main function, wrapping the one we provide inside of another function that does some initializing and redirecting of the standard buffers.

The first (quite hackish solution) is to simply short-circuit the SDL2main library and  all of its redefinitions by adding the following in the main.cpp file:

```c++
#ifdef main
#undef main
#endif
```

It works, but since I didn't really like this solution, I dug a bit deeper and found out some more problems:

1. The SDL_main.lib file is not properly added in the dependencies by our CMake build configuration: In `cmake/FindSDL2.cmake`, in the `FIND_LIBRARY` calls, the `NAMES` argument is trying to find ONE of the names provided. Providing several names does not result in several libraries being added. (https://stackoverflow.com/a/15212241/3909725)
This can be fixed by the following snippets of code:
```cmake
# cmake/FindSDL2.cmake 
FIND_LIBRARY(SDL2MAIN_LIBRARY
	NAMES
		SDL2main
	PATHS
		${SDL2_SEARCH_PATHS}
	PATH_SUFFIXES
		lib
		lib64
		lib/x86
		lib/x64
	DOC
		"The SDL2main library"
)

# CMakeLists.txt
target_link_libraries( CGFX5
	${OPENGL_LIBRARIES}
	${GLEW_LIBRARIES}
	${SDL2_LIBRARY}
	${SDL2MAIN_LIBRARY}
	${ASSIMP_LIBRARIES}
)

```
Please note that **the problem is also present for GLEW**, but strangely enough, the fact that for example `glew32s.lib` is not linked does not seem to break my build.

2. Now that SDL2main.lib is correctly added to the linker configuration, I encountered a new error!

![link_error_2](https://user-images.githubusercontent.com/6213717/39131366-8a2cd8b2-470f-11e8-9ef5-3459581fe43b.JPG)

| Severity |   Code  |                                  Description                                   | Project |                                  File                                  | Line | Suppression State |
|----------|---------|--------------------------------------------------------------------------------|---------|------------------------------------------------------------------------|------|-------------------|
| Error    | LNK2019 | unresolved external symbol __imp____iob_func referenced in function _ShowError | CGFX5   | C:\Users\thundzz\Desktop\CGFX5\build\SDL2main.lib(SDL_windows_main.obj) |    1 |                   |
| Error    | LNK1120 | 1 unresolved externals                                                         | CGFX5   | C:\Users\thundzz\Desktop\CGFX5\build\Debug\CGFX5.exe                    |    1 |                   |

Fortunately, after a quick googling of the problem, I found the reason for it ! [Visual Studio 2015 changed the way `stdin`, `stdout` and `stderr` are defined](https://stackoverflow.com/questions/30412951/unresolved-external-symbol-imp-fprintf-and-imp-iob-func-sdl2) ...
Now... the clean solution would be to build SDL2 with VS2015 and link the output .obj file instead of the one currently provided, but that would break the build for  users of previous versions.

The other solution (which was proposed by someone on the stackoverflow thread above) is to redeclare the previously used symbols:
```c++
extern "C" { FILE __iob_func[3] = { *stdin,*stdout,*stderr }; }
```

I don't really like this solution either, but it looks to me like it is the _least worst_ solution of solving the issue.


Sorry for dumping such huge walls of text, but I felt the need to explain how I came to these conclusions instead of just throwing the final thing at you.
I thought I'd start by sharing all of this before trying to propose any new pull request. If you want me to commit the changes on another branch and submit a new pull request, please tell me!

BennyQBD commented:
Thanks again for being so thorough! It's really helpful, and has made this much easier to resolve. As far as the platform include, easy fix was just an oversight on my part as you suspected.

As for the libraries, that's an impressive depth of understanding you've developed for the project's dependencies, and I'm pleased you've discovered many things that could be a big issue if they weren't intentional. However, as odd as this may seem, what you're describing is actually by design. glew32s.lib isn't supposed to be linked, because it's the same as glew32.lib, just a static version. Since the project is already using the dynamically linked glew32.lib, there's no need do "double link" a second version of glew. None of the extra SDL libraries are linked because the only thing that *should* be needed is the primary SDL2 library. Which once again brings us back to the bizarre issue with the main function.

I've read through the SDL2 source code on this part. Here's what I've discovered:

- On all platforms except Windows and game consoles (i.e. phones, Linux, and Mac), the main redefinition does nothing.

- On Windows, if you're building a command line app (i.e. what this is doing), it also does nothing. However, if you're building a windowed app using the Windows API, then it does some sorcery to get the "command line parameters" and tries to *pretend* you're running a command line app.

- On game consoles, it tries to do whatever is necessary to appease the game console's application requirements and then still *pretend* it's actually a command line app.

I think it's notable that in all 3 cases, what SDL is doing is NOT what you want.

- If you're deliberately making a command line app, there's no reason to have that wrapped in a needless function. Thus, SDL2's wrapper is undesirable.

- If you're deliberately using the Windows API, then obviously, you're NOT trying to make a command line app, so you probably shouldn't be pretending it's a command line app. Thus, SDL2's wrapper is undesirable.

- If you're making an app for the game console, then all the console-specific technology is part of the appeal, and you'll almost certainly need to take advantage of it to some point to get your program working as expected. Silently appeasing and ignoring it is probably not what you want, and you almost certainly don't want to pretend it's another command line app. Thus, SDL2's wrapper is undesirable.

Given this, as hackish as it may appear at first, I believe that undefing main is the correct solution. What you really want is control over your programs entry point, and what SDL does with it simply isn't useful as far as I can tell. So I've added the undef main trick as the solution.

If you could give it one more run to see if all works as expected, I'd appreciate it! If no more issues crop up, I think we can safely close this. Thanks again for all your help, it's much appreciated!

Thundzz commented:
Thank you for all the explanations! Now that I understand better, it sounds like the solution you implemented is actually the most relevant for this use case.

I just re-cloned the project and now everything builds and runs smoothly!
As you said, we can now safely close this... so I'm doing that right now. 😄
